R/array.R

```diff
@@ -44,3 +44,47 @@ np_array <- function(data, dtype = NULL, order = "C") {
     as_r_value(x$size)
 }
 
+#' Reshape an Array
+#' 
+#' Reshape (reindex) a multi-dimensional array, using reshaping semantics that
+#' align with the behavior of NumPy's own `reshape()` function.
+#' 
+#' This function differs from e.g. `dim(x) <- dim` in a very important way: by
+#' default, `array_reshape()` will fill the new dimensions in row-major (`C`-style)
+#' ordering, while [dim<-()] will fill new dimensions in column-major
+#' (`F`ortran-style) ordering. This is done to be consistent with the NumPy
+#' `np.reshape()` function, which defaults to this sort of ordering when
+#' reshaping arrays. See the examples for why this difference may be important.
+#' 
+#' @param x Either an \R array or a NumPy array.
+#' @param dim The new dimensions to be set on the array.
+#' @param order The order in which elements of `x` should be read during
+#'   the rearrangement. `"C"` means elements should be read in row-major
+#'   order, with the last index changing fastest; `"F"` means elements should
+#'   be read in column-major order, with the first index changing fastest.
+#' 
+#' @examples
+#' # let's construct a 2x2 array from a vector of 4 elements
+#' x <- 1:4
+#' 
+#' # rearrange will fill the array row-wise
+#' array_reshape(x, c(2, 2))
+#' #      [,1] [,2]
+#' # [1,]    1    2
+#' # [2,]    3    4
+#' # setting the dimensions 'fills' the array col-wise
+#' dim(x) <- c(2, 2)
+#' x
+#' #      [,1] [,2]
+#' # [1,]    1    3
+#' # [2,]    2    4
+#' @export
+array_reshape <- function(x, dim, order = c("C", "F")) {
+  np <- import("numpy", convert = FALSE)
+  order <- match.arg(order)
+  reshaped <- np$reshape(x, as.integer(dim), order)
```

Comment:
At one point had you speculated that the `order` argument affected how the inbound memory was interpreted? I wasn't sure whether that was in fact the case, but if it is then it seems like `order = "C"` would be wrong for R arrays. Let me know what you think.

Reply:
This bit is definitely a bit confusing, but I think this is still correct.

The 'order' argument controls how, exactly, the data is reshaped / re-indexed; in particular, when increasing dimensionality, it controls whether the last dimension grows fastest (C-style) or the first (Fortran, or F-style). For example, if you had the vector 1:4, reshaping it to a 2x2 matrix in a C-style way would imply filling that new matrix by rows, then by columns, e.g.

```R
> array_reshape(1:4, dim = c(2, 2))
     [,1] [,2]
[1,]    1    2
[2,]    3    4
```

It's worth noting that doing C-style reshaping on an F-array implies that the underlying data buffer will be mutated; that is, we're doing more than just adjusting the dimensions, we're also reshaping the data so the new indexing scheme can be accommodated. It's also worth noting that this is not wrong; it's just a consequence of doing this C-style reshaping of an F-array.

```R
> c(array_reshape(1:4, dim = c(2, 2)))
[1] 1 3 2 4
```

It does imply that this form of reshaping will be inefficient for R arrays, but that's the cost we pay for correctness here.

Reply:
Okay, sounds good! Just wanted to make sure.